lib/src/main/kotlin/com/swmansion/starknet/provider/gateway/GatewayProvider.kt

```diff
@@ -99,12 +96,19 @@ class GatewayProvider(
         return response.body
     }
 
-    private fun <T> buildDeserializer(deserializationStrategy: DeserializationStrategy<T>): HttpResponseDeserializer<T> =
+    private fun <T> buildDeserializer(
+        deserializationStrategy: DeserializationStrategy<T>,
+        deserializationExceptionHandler: ((String) -> Exception?)? = null,
+    ): HttpResponseDeserializer<T> =
         Function { response ->
             val body = handleResponseError(response)
 
             val json = Json { ignoreUnknownKeys = true }
-            json.decodeFromString(deserializationStrategy, body)
+            try {
+                json.decodeFromString(deserializationStrategy, body)
+            } catch (e: SerializationException) {
+                throw deserializationExceptionHandler?.invoke(body) ?: e
+            }
         }
```

Comment:
So honestly I don't know if this is good code, perhaps it's not. Problem is, in Gateway, when querying for transaction that was never received, a 200 OK is returned with JSON

```json
{
    "status": "NOT_RECEIVED"
}
```

This has to be handled somewhere since the deserialization will fail as there's no data class for transaction like this defined obviously.

I decided to add a fallback to deserializer that allows for custom error handling in `SerializationException` is thrown, but I'm unsure if this is a good flow.

Another approach I've considered is having this logic defined somewhere in GatewayTransactionTransformingSerializer (or some entirely custom GatewayTransactionSerializer).